In [19]:
from IPython.display import display, HTML

display(HTML("""
<style>
    /* 1. UMUMIY KONTEYNER VA SHRIFT */
    .rendered_html, .jp-RenderedHTMLCommon, .jp-RenderedMarkdown {
        font-family: "Times New Roman", Times, serif !important;
        color: black !important;
        font-feature-settings: "kern" 1, "liga" 1, "clig" 1, "dlig" 1 !important;
        -webkit-font-smoothing: antialiased !important;
    }

    /* 2. MATNLAR: 1.5 oraliq, abzats va ro‘yxatlar */
    .rendered_html p, .jp-RenderedHTMLCommon p {
        font-family: "Times New Roman", Times, serif !important;
        font-size: 14pt !important;
        line-height: 1.5 !important;
        text-align: justify !important;
        text-indent: 1.25cm !important; /* Faqat p uchun abzats */
        margin: 0 !important;
    }

    /* Ro‘yxatlar uchun maxsus (nuqtalarni qaytarish) */
    .rendered_html ul, .jp-RenderedHTMLCommon ul {
        list-style-type: disc !important;
        margin-left: 2.5cm !important; /* Nuqtalar ko‘rinishi uchun surish */
        padding: 0 !important;
        text-indent: 0 !important;
    }

    .rendered_html li, .jp-RenderedHTMLCommon li {
        display: list-item !important; /* Block-dan list-item-ga qaytarish */
        font-family: "Times New Roman", Times, serif !important;
        font-size: 14pt !important;
        line-height: 1.5 !important;
        text-indent: 0 !important;
        margin: 0 !important;
    }

    .rendered_html p, .jp-RenderedHTMLCommon p {
        text-indent: 1.25cm !important;
    }

    /* 3. SARLAVHALAR: Markazda */
    .rendered_html h1, .jp-RenderedHTMLCommon h1,
    .rendered_html h2, .jp-RenderedHTMLCommon h2,
    .rendered_html h3, .jp-RenderedHTMLCommon h3 {
        font-family: "Times New Roman", Times, serif !important;
        text-align: center !important;
        text-indent: 0 !important;
        margin: 0.5em 0 !important;
        font-weight: bold !important;
    }
    
    /* 4. KOD BLOKLARI VA PROMPTLARNI (In/Out) YASHIRISH */
    :root {
        --jp-code-font-size: 12pt !important;
    }

    /* "In [ ]" va "Out [ ]" yozuvlarini butunlay yo‘qotish */
    .jp-InputPrompt, .jp-OutputPrompt, .prompt, .input_prompt, .output_prompt {
        display: none !important;
        width: 0 !important;
        visibility: hidden !important;
    }

    /* Kod bloklarini chap chekkaga taqash */
    .jp-CodeCell .jp-Cell-inputWrapper, .cell.code_cell .input {
        margin-left: 0 !important;
        padding-left: 0 !important;
    }

    /* Sliderni yo‘qotish va qatorni majburan sindirish */
    .jp-CodeMirror, .CodeMirror, .CodeMirror-scroll, .CodeMirror-sizer, .CodeMirror-lines,
    .jp-OutputArea-output pre, .output_area pre, 
    .CodeMirror-code, .CodeMirror-line,
    code, kbd, samp {
        font-family: "Consolas", monospace !important;
        font-size: 12pt !important;
        text-indent: 0 !important;
        line-height: 1.3 !important;
        white-space: pre-wrap !important;
        word-wrap: break-word !important;
        word-break: break-all !important;
        overflow-x: visible !important;
        height: auto !important;
    }

    /* 5. BIRINCHI KATAKCHANI YASHIRISH VA TANLANGANDA OCHISH */
    /* Odatiy holatda mutlaqo yashirin va umuman joy egallamaydi */
    .jp-Notebook .jp-Cell:first-child:not(.jp-mod-selected), 
    .container div.cell.code_cell:first-of-type:not(.selected) {
        position: absolute !important;
        height: 0 !important;
        min-height: 0 !important;
        width: 0 !important;
        margin: 0 !important;
        padding: 0 !important;
        border: none !important;
        overflow: hidden !important;
        opacity: 0 !important;
        pointer-events: none !important;
        visibility: hidden !important;
    }

    

    /* Katakcha tanlanganda (Ctrl+A bosilganda) to‘liq ochiladi */
    .jp-Notebook .jp-Cell:first-child.jp-mod-selected,
    .container div.cell.code_cell:first-of-type.selected {
        position: relative !important;
        height: auto !important;
        width: 100% !important;
        opacity: 1 !important;
        overflow: visible !important;
        pointer-events: auto !important;
        visibility: visible !important;
    }
    
    /* Jadvallar ichidagi matnlarni 14pt qilish */
    .rendered_html table, .jp-RenderedHTMLCommon table,
    .rendered_html td, .jp-RenderedHTMLCommon td,
    .rendered_html th, .jp-RenderedHTMLCommon th {
        font-family: "Times New Roman", Times, serif !important;
        font-size: 14pt !important;
        line-height: 1.5 !important;
        color: black !important;
    }
</style>
"""))


## 1.3. NumPy massivlarida amallar (Operations in NumPy Array)

NumPy massivlari ma'lumotlar bilan samarali ishlash (data manipulation) imkonini beruvchi to'rtta asosiy amal turini taqdim etadi:

* **Binar amallar (Binary Operations)** – Elementlar bo'yicha solishtirish (masalan, kattaroq, teng) va "VA" (AND) yoki "YOKI" (OR) kabi mantiqiy operatsiyalarni bajaradi.
* **Matematik amallar (Mathematical Operations)** – `sum()` (yig'indi), `mean()` (o'rtacha qiymat) va `sqrt()` (kvadrat ildiz) kabi funksiyalar tezkor va samarali sonli hisob-kitoblarni amalga oshirish imkonini beradi.
* **Satr amallari (String Operations)** – Massivlar ichidagi matnlarni birlashtirish (concatenation), bosh harflarga o'tkazish (uppercasing) va almashtirish (replacing) kabi amallar uchun `np.char` funksiyalaridan foydalaniladi.
* **Arifmetik amallar (Arithmetic Operations)** – Massivlarda elementlar bo'yicha qo'shish, ayirish, ko'paytirish va bo'lish amallarini tezkor bajarishni ta'minlaydi.

### 1.3.1. Binar amallar (Binary Operations)

**Binar amallar (Binary operations)** bitlar ustida ishlaydi va bitma-bit operatsiyalarni amalga oshiradi. Oddiy qilib aytganda, binar amal — bu yangi qiymat hosil qilish uchun ikkita qiymatni birlashtirish qoidasidir.

**`numpy.bitwise_and()`**

Ushbu funksiya ikkita massiv elementlarining bitma-bit **"VA" (AND)** amalini hisoblash uchun ishlatiladi. Funksiya kirish massivlaridagi butun sonlarning asosiy binar ko'rinishi (ikkilik sanoq tizimidagi ko'rinishi) ustida amallarni bajaradi.

**1-misol: Oddiy sonlar ustida qo'llash**

In [3]:
import numpy as np

# Kirish sonlari
son1 = 10  # Binar ko'rinishi: 1010
son2 = 11  # Binar ko'rinishi: 1011

print("1-son:", son1)
print("2-son:", son2)

# Bitma-bit AND amali
natija = np.bitwise_and(son1, son2)
print("10 va 11 ning bitwise_and natijasi:", natija)

1-son: 10
2-son: 11
10 va 11 ning bitwise_and natijasi: 10


**2-misol: Massivlar ustida qo'llash**

In [4]:
# Kirish massivlari
massiv1 = [2, 8, 125]
massiv2 = [3, 3, 115]

print("1-massiv:", massiv1)
print("2-massiv:", massiv2)

# Elementlar bo'yicha bitma-bit AND amali
natija_massiv = np.bitwise_and(massiv1, massiv2)
print("bitwise_and natijasi:", natija_massiv)

1-massiv: [2, 8, 125]
2-massiv: [3, 3, 115]
bitwise_and natijasi: [  2   0 113]


**Izoh:**
Funksiya har bir elementni o'zaro solishtiradi. Masalan, ikkinchi juftlik elementlar uchun:
* `8` ning binar shakli: `1000`
* `3` ning binar shakli: `0011`
* Ular o'rtasida **AND** amali bajarilganda hech qanday bit mos kelmaydi, shuning uchun natija `0` bo'ladi.

**`numpy.bitwise_or()`**

Ushbu funksiya ikkita massiv elementlarining bitma-bit **"YOKI" (OR)** amalini hisoblash uchun ishlatiladi. Funksiya kirish massivlaridagi butun sonlarning asosiy binar (ikkilik) ko'rinishi ustida amallarni bajaradi.

**1-misol: Oddiy sonlar ustida qo'llash**

In [5]:
# Kirish sonlari
son1 = 10  # Binar ko'rinishi: 1010
son2 = 11  # Binar ko'rinishi: 1011

print("1-son:", son1)
print("2-son:", son2)

# Bitma-bit OR amali
natija = np.bitwise_or(son1, son2)
print("10 va 11 ning bitwise_or natijasi:", natija)

1-son: 10
2-son: 11
10 va 11 ning bitwise_or natijasi: 11


**2-misol: Massivlar ustida qo'llash**

In [6]:
# Kirish massivlari
massiv1 = [2, 8, 125]
massiv2 = [3, 3, 115]

print("1-massiv:", massiv1)
print("2-massiv:", massiv2)

# Elementlar bo'yicha bitma-bit OR amali
natija_massiv = np.bitwise_or(massiv1, massiv2)
print("bitwise_or natijasi:", natija_massiv)

1-massiv: [2, 8, 125]
2-massiv: [3, 3, 115]
bitwise_or natijasi: [  3  11 127]


### Mantiqiy tahlil: Nima uchun 8 va 3 dan 11 hosil bo‘ldi?

Bitma-bit **OR (YOKI)** amali bajarilishi uchun solishtirilayotgan bitlarning kamida bittasi `1` bo‘lishi kifoya.



$$
\begin{array}{r@{\quad}l}
8 = & 1000_2 \\
3 = & 0011_2 \\
\hline
\text{OR} = & 1011_2 \rightarrow (11_{10})
\end{array}
$$

**Tushuntirish:**
1. Oxirgi bitlar: `0` va `1` $\rightarrow$ natija `1`.
2. Undan oldingi: `0` va `1` $\rightarrow$ natija `1`.
3. Undan oldingi: `0` va `0` $\rightarrow$ natija `0`.
4. Birinchi bitlar: `1` va `0` $\rightarrow$ natija `1`.

Natijada `1011` hosil bo'ladi, bu o'nlik sanoq tizimida **11** ga teng.

**`numpy.bitwise_xor()`**

Ushbu funksiya ikkita massiv elementlarining bitma-bit **"istisno qiluvchi YOKI" (XOR)** amalini hisoblash uchun ishlatiladi. Funksiya kirish massivlaridagi butun sonlarning asosiy binar (ikkilik) ko'rinishi ustida amallarni bajaradi.

**1-misol: Oddiy sonlar ustida qo'llash**

In [7]:
# Kirish sonlari
son1 = 10  # Binar ko'rinishi: 1010
son2 = 11  # Binar ko'rinishi: 1011

print("1-son:", son1)
print("2-son:", son2)

# Bitma-bit XOR amali
natija = np.bitwise_xor(son1, son2)
print("10 va 11 ning bitwise_xor natijasi:", natija)

1-son: 10
2-son: 11
10 va 11 ning bitwise_xor natijasi: 1


**2-misol: Massivlar ustida qo'llash**

In [8]:
# Kirish massivlari
massiv1 = [2, 8, 125]
massiv2 = [3, 3, 115]

print("1-massiv:", massiv1)
print("2-massiv:", massiv2)

# Elementlar bo'yicha bitma-bit XOR amali
natija_massiv = np.bitwise_xor(massiv1, massiv2)
print("bitwise_xor natijasi:", natija_massiv)

1-massiv: [2, 8, 125]
2-massiv: [3, 3, 115]
bitwise_xor natijasi: [ 1 11 14]


### Mantiqiy tahlil: XOR qanday ishlaydi?

**XOR (Exclusive OR)** amali "farqli bo'lsa 1, bir xil bo'lsa 0" qoidasi asosida ishlaydi. Ya'ni, solishtirilayotgan bitlar har xil bo'lgandagina natija `1` bo'ladi.



**Misol: 8 va 3 dan nima uchun 11 hosil bo'ldi?**
$$
\begin{array}{r@{\quad}l}
8 = & 1000_2 \\
3 = & 0011_2 \\
\hline
\text{XOR} = & 1011_2 \rightarrow (11_{10})
\end{array}
$$

**Misol: 125 va 115 dan nima uchun 14 hosil bo'ldi?**
$$
\begin{array}{r@{\quad}l}
125 = & 1111101_2 \\
115 = & 1110011_2 \\
\hline
\text{XOR} = & 0001110_2 \rightarrow (14_{10})
\end{array}
$$

**Tushuntirish:**
* Ikkala bit `1` bo'lsa $\rightarrow$ `0`
* Ikkala bit `0` bo'lsa $\rightarrow$ `0`
* Bittasi `1` va bittasi `0` bo'lsa $\rightarrow$ `1`

**`numpy.invert()`**

Ushbu funksiya massiv elementlarining bitma-bit **teskarisini (Inversion)** hisoblash uchun ishlatiladi. U kirish massivlaridagi butun sonlarning binar ko'rinishi ustida **"EMAS" (NOT)** mantiqiy amalini bajaradi.

**Muhim jihati:**
Ishorali butun sonlar (signed integers) uchun funksiya **ikki qo'shimcha (two’s complement)** qiymatni qaytaradi. Ikki qo'shimcha tizimida manfiy sonlar mutlaq qiymatning (absolute value) teskarisiga birni qo'shish orqali ifodalanadi. Oddiy qilib aytganda, butun sonlar uchun natija $-(x + 1)$ formulasiga muvofiq bo'ladi.

#### **1-misol: Oddiy son ustida qo'llash**

In [9]:
in_num = 10
print("Kirish soni:", in_num)

# Bitma-bit NOT amali
out_num = np.invert(in_num)
print("10 ning inversiyasi:", out_num)

Kirish soni: 10
10 ning inversiyasi: -11


#### **2-misol: Massiv ustida qo'llash**

In [10]:
in_arr = [2, 0, 25]
print("Kirish massivi:", in_arr)

# Massiv elementlari uchun inversiya
out_arr = np.invert(in_arr)
print("Inversiyadan keyingi natija:", out_arr)

Kirish massivi: [2, 0, 25]
Inversiyadan keyingi natija: [ -3  -1 -26]


### **Mantiqiy tahlil: Nima uchun 10 dan -11 hosil bo'ldi?**

Kompyuterda ishorali sonlar xotirada qanday saqlanishini tushunish natijani izohlashga yordam beradi:

1.  **10 sonining binar ko'rinishi (8-bitda):** `00001010`
2.  **Bitma-bit invert (NOT) amali:** `11110101`
3.  **Ikki qo'shimcha tizimida:** `11110101` soni manfiy sonni anglatadi. Uni o'nlik tizimga qaytarish uchun barcha bitlar teskarisiga aylantiriladi (`00001010`) va `1` qo'shiladi (`00001011`). Bu esa **11** ga teng, ishorasi manfiy bo'lgani uchun natija **-11** bo'ladi.



**Qisqa formula:**
> Har qanday butun $x$ soni uchun `np.invert(x)` natijasi **$-(x + 1)$** ga teng bo'ladi.
> * Masalan: $2 \rightarrow -(2 + 1) = -3$
> * Masalan: $0 \rightarrow -(0 + 1) = -1$
> * Masalan: $25 \rightarrow -(25 + 1) = -26$

**`numpy.left_shift()`**

Ushbu funksiya butun sonning bitlarini chapga siljitish uchun ishlatiladi. Siljitish jarayoni birinchi massivdagi (`arr1`) sonlarning o'ng tomoniga ikkinchi massivda (`arr2`) ko'rsatilgan miqdorda nollarni qo'shish orqali amalga oshiriladi.



Sonlarning ichki ko'rinishi binar (ikkilik) formatda bo'lgani uchun, bu amal matematik jihatdan birinchi sonni $2^{arr2}$ ga ko'paytirish bilan barobardir. Masalan, agar son **5** bo'lsa va biz uni **2 bit** chapga siljitmoqchi bo'lsak, natija $5 \cdot 2^2 = 20$ bo'ladi.

#### **1-misol: Oddiy son ustida qo'llash**

In [11]:
in_num = 5
bit_shift = 2

print("Kirish soni:", in_num)
print("Siljitish kerak bo'lgan bitlar soni:", bit_shift)

# Chapga siljitish amali
out_num = np.left_shift(in_num, bit_shift)
print("2 bit chapga siljitgandan keyin:", out_num)

Kirish soni: 5
Siljitish kerak bo'lgan bitlar soni: 2
2 bit chapga siljitgandan keyin: 20


#### **2-misol: Massiv ustida qo'llash**

In [12]:
in_arr = [2, 8, 15]
bit_shift = [3, 4, 5]

print("Kirish massivi:", in_arr)
print("Siljitish miqdori:", bit_shift)

# Massiv elementlarini mos ravishda siljitish
out_arr = np.left_shift(in_arr, bit_shift)
print("Natija massivi:", out_arr)

Kirish massivi: [2, 8, 15]
Siljitish miqdori: [3, 4, 5]
Natija massivi: [ 16 128 480]


### **Mantiqiy tahlil: Nima uchun 5 dan 20 hosil bo'ldi?**

Keling, buni binar (ikkilik) tizimda ko'rib chiqamiz:

1.  **5** sonining binar shakli: `00000101`
2.  **2 bit chapga siljitish:** Bitlarni chapga suramiz va o'ngdan ikkita `0` qo'shamiz: `00010100`
3.  **Hosil bo'lgan son:** `00010100` binar tizimda **20** ga teng.

**Matematik tekshirish:**
* $2 \cdot 2^3 = 2 \cdot 8 = 16$
* $8 \cdot 2^4 = 8 \cdot 16 = 128$
* $15 \cdot 2^5 = 15 \cdot 32 = 480$

Chapga siljitish amali asosan tezkor ko'paytirish sifatida dasturlashda keng qo'llaniladi.

**`numpy.right_shift()`**

Ushbu funksiya butun sonning bitlarini o'ngga siljitish uchun ishlatiladi. Kompyuterda sonlarning ichki ko'rinishi binar (ikkilik) formatda bo'lgani sababli, bu amal matematik jihatdan birinchi sonni (**arr1**) $2^{arr2}$ ga bo'lish (butun qismini olish) bilan barobardir.



Masalan, agar son **20** bo'lsa va biz uni **2 bit** o'ngga siljitmoqchi bo'lsak, natija $20 / 2^2 = 5$ bo'ladi.

#### **1-misol: Oddiy son ustida qo'llash**

In [13]:
in_num = 20
bit_shift = 2

print("Kirish soni:", in_num)
print("Siljitish kerak bo'lgan bitlar soni:", bit_shift)

# O'ngga siljitish amali
out_num = np.right_shift(in_num, bit_shift)
print("2 bit o'ngga siljitgandan keyin:", out_num)

Kirish soni: 20
Siljitish kerak bo'lgan bitlar soni: 2
2 bit o'ngga siljitgandan keyin: 5


#### **2-misol: Massiv ustida qo'llash**

In [14]:
in_arr = [24, 48, 16]
bit_shift = [3, 4, 2]

print("Kirish massivi:", in_arr)
print("Siljitish miqdori:", bit_shift)

# Massiv elementlarini mos ravishda o'ngga siljitish
out_arr = np.right_shift(in_arr, bit_shift)
print("Natija massivi:", out_arr)

Kirish massivi: [24, 48, 16]
Siljitish miqdori: [3, 4, 2]
Natija massivi: [3 3 4]


### **Mantiqiy tahlil: Nima uchun 20 dan 5 hosil bo'ldi?**

Binar (ikkilik) tizimda jarayon quyidagicha ko'rinadi:

1.  **20** sonining binar shakli: `00010100`
2.  **2 bit o'ngga siljitish:** Barcha bitlarni 2 marta o'ngga suramiz. O'ng chekkadagi bitlar "tushib qoladi", chapdan esa nollar qo'shiladi: `00000101`
3.  **Hosil bo'lgan son:** `00000101` binar tizimda **5** ga teng.

**Matematik tekshirish:**
* $24 / 2^3 = 24 / 8 = 3$
* $48 / 2^4 = 48 / 16 = 3$
* $16 / 2^2 = 16 / 4 = 4$

O'ngga siljitish amali dasturlashda sonni 2 ning darajalariga juda tez (optimizatsiya qilingan tarzda) bo'lish uchun ishlatiladi.

 **`numpy.binary_repr(son, width=None)`**

Ushbu funksiya kiritilgan sonning ikkilik (binar) ko‘rinishini **satr (string)** shaklida qaytarish uchun ishlatiladi.

**Manfiy sonlar bilan ishlash xususiyati:**
* Agar `width` (kenglik) parametri berilmasa, manfiy sonning oldiga oddiygina minus ("-") belgisi qo‘yiladi.
* Agar `width` parametri ko‘rsatilsa, sonning shu kenglikka nisbatan **ikki qo‘shimcha (two’s complement)** kodi qaytariladi.



Kompyuterlarda ishorali butun sonlarni ifodalashning eng keng tarqalgan usuli aynan ikki qo‘shimcha tizimidir. Bunda manfiy son mutlaq qiymatning ikkilikdagi teskarisiga birni qo‘shish orqali hosil qilinadi.

#### **1-misol: Oddiy sonning ikkilik ko‘rinishi**

In [15]:
son = 10
print("Kirish soni:", son)

# 10 sonining ikkilik ko'rinishini olamiz
natija = np.binary_repr(son)
print("10 ning ikkilik ko'rinishi:", natija)

Kirish soni: 10
10 ning ikkilik ko'rinishi: 1010


#### **2-misol: width parametri bilan ishlash (Musbat va manfiy sonlar)**

In [16]:
# Musbat 5 va manfiy 8 sonlari
sonlar = [5, -8]

# 5 soni uchun (width ishlatmasdan va ishlatib)
print("5 sonining ikkilik ko'rinishi:")
print("width ishlatilmasa:", np.binary_repr(sonlar[0])) 
print("width = 5 bo'lganda:", np.binary_repr(sonlar[0], width=5)) 

print("\n-8 sonining ikkilik ko'rinishi:")
# -8 uchun (width ishlatilmasa va ishlatib)
print("width ishlatilmasa:", np.binary_repr(sonlar[1])) 
print("width = 5 bo'lganda (ikki qo'shimcha kodi):", np.binary_repr(sonlar[1], width=5))

5 sonining ikkilik ko'rinishi:
width ishlatilmasa: 101
width = 5 bo'lganda: 00101

-8 sonining ikkilik ko'rinishi:
width ishlatilmasa: -1000
width = 5 bo'lganda (ikki qo'shimcha kodi): 11000


### **Mantiqiy tahlil: Nima uchun -8 soni '11000' bo‘ldi?**

`width=5` bo‘lganda, manfiy son ikki qo‘shimcha tizimi bo‘yicha hisoblanadi:
1.  **8** sonining 5-bitli ko'rinishi: `01000`
2.  **Inversiya (teskarisiga aylantirish):** `10111`
3.  **1 qo'shish:** `10111 + 1 = 11000`

Bu usul dasturchilarga sonlarning xotiradagi haqiqiy bit ko‘rinishini satr sifatida ko‘rish imkonini beradi.

**`numpy.packbits(myarray, axis=None)`**

Ushbu funksiya binar qiymatlardan (0 va 1) tashkil topgan massiv elementlarini bitlar sifatida **uint8** (8-bitli ishorasiz butun son) turidagi massivga joylashtiradi (paketlaydi). Agar elementlar soni to‘liq baytni (8 bitni) tashkil qilmasa, natija oxiriga nol bitlar qo‘shish orqali to‘ldiriladi.

#### **Misol:**

In [17]:
# 3 o'lchamli binar massiv yaratamiz
a = np.array([[[1, 0, 1],
               [0, 1, 0]],
              [[1, 1, 0],
               [0, 0, 1]]])

# Oxirgi o'q (axis=-1) bo'yicha elementlarni bitlarga joylashtiramiz
b = np.packbits(a, axis=-1)

print(b)

[[[160]
  [ 64]]

 [[192]
  [ 32]]]


### **Mantiqiy tahlil: Natijalar qayerdan keldi?**

Ushbu funksiya qanday ishlashini tushunish uchun har bir kichik massivni ko‘rib chiqamiz. `axis=-1` bo‘lgani uchun u eng ichki qavslardagi 3 ta elementni oladi va ularni 8-bitli baytning boshiga qo‘yadi, qolgan 5 ta bo‘sh joyni esa nollar bilan to‘ldiradi.

**1. Birinchi element: `[1, 0, 1]`**
* Bitlar: `1 0 1` + (5 ta nol) `0 0 0 0 0` = `10100000`
* Ikkilikdan o'nlikka: $2^7 + 2^5 = 128 + 32 = \mathbf{160}$

**2. Ikkinchi element: `[0, 1, 0]`**
* Bitlar: `0 1 0` + (5 ta nol) `0 0 0 0 0` = `01000000`
* Ikkilikdan o'nlikka: $2^6 = \mathbf{64}$

**3. Uchinchi element: `[1, 1, 0]`**
* Bitlar: `1 1 0` + (5 ta nol) `0 0 0 0 0` = `11000000`
* Ikkilikdan o'nlikka: $2^7 + 2^6 = 128 + 64 = \mathbf{192}$

**4. To'rtinchi element: `[0, 0, 1]`**
* Bitlar: `0 0 1` + (5 ta nol) `0 0 0 0 0` = `00100000`
* Ikkilikdan o'nlikka: $2^5 = \mathbf{32}$

> **Xulosa:** `packbits` funksiyasi binar ma'lumotlarni xotirada 8 barobar ixchamroq saqlash imkonini beradi, chunki har bir 0 yoki 1 qiymati alohida bayt emas, balki baytning bitta biti sifatida saqlanadi.

**`numpy.unpackbits(myarray, axis=None)`**

Ushbu funksiya `uint8` (8-bitli ishorasiz butun son) turidagi massiv elementlarini qismlarga ajratib, ularni binar qiymatli (0 va 1 lardan iborat) massivga aylantiradi. `myarray` ning har bir elementi bitta bit maydoni deb qaraladi va u 8 ta alohida bitga yoyib chiqiladi.

**Natijaviy massiv shakli:**
* Agar `axis=None` bo'lsa, natija 1 o'lchamli (1-D) massiv bo'ladi.
* Agar o'q (`axis`) ko'rsatilsa, natijaviy massiv kirish massivi bilan bir xil shaklda bo'ladi, faqat ko'rsatilgan o'q bo'yicha bitlar yoyiladi.

#### **Misol:**

In [18]:
# uint8 turidagi ustun ko'rinishidagi massiv yaratamiz
a = np.array([[2], [7], [23]], dtype=np.uint8)

# Elementlarni 1-o'q (ustunlar yo'nalishi) bo'yicha binar ko'rinishga yoyamiz
b = np.unpackbits(a, axis = 1)

print(b)

[[0 0 0 0 0 0 1 0]
 [0 0 0 0 0 1 1 1]
 [0 0 0 1 0 1 1 1]]


### **Mantiqiy tahlil: Natijalar qanday hosil bo'ldi?**

Har bir butun son 8-bitli xotira katakchasida qanday saqlangan bo'lsa, `unpackbits` uni aynan shunday ko'rinishda massiv elementlariga ajratadi:

1.  **2 soni uchun:**
    * 2 ning ikkilik sanoq tizimidagi ko'rinishi: `10`
    * 8-bitli formatda (chapdan nollar bilan to'ldirilgan): `00000010`
    * Natija: `[0, 0, 0, 0, 0, 0, 1, 0]`

2.  **7 soni uchun:**
    * 7 ning ikkilik ko'rinishi: `111`
    * 8-bitli formatda: `00000111`
    * Natija: `[0, 0, 0, 0, 0, 1, 1, 1]`

3.  **23 soni uchun:**
    * 23 ning ikkilik ko'rinishi: $16 + 4 + 2 + 1 = 10111$
    * 8-bitli formatda: `00010111`
    * Natija: `[0, 0, 0, 1, 0, 1, 1, 1]`

> **Xulosa:** `unpackbits` funksiyasi `packbits` funksiyasining teskarisi bo'lib, xotirada siqilgan holatdagi bitlarni qaytadan tahlil qilish (masalan, tasvirlarni qayta ishlash yoki binar ma'lumotlarni o'qish) uchun juda qulaydir.

### 1.3.2. Matematik amallar (Mathematical Operations)

NumPy kutubxonasi matematik amallarning ulkan to'plamini o'z ichiga oladi. U standart trigonometrik funksiyalar, arifmetik amallar, kompleks sonlar bilan ishlash va boshqa ko'plab imkoniyatlarni taqdim etadi.

### Trigonometrik funksiyalar

NumPy radianlarda berilgan burchaklar uchun trigonometrik nisbatlarni elementlar bo'yicha (element-wise) hisoblash uchun `sin()`, `cos()` va `tan()` kabi funksiyalarni taqdim etadi.



Quyidagi jadvalda NumPy-dagi asosiy trigonometrik funksiyalar va ularning tavsifi keltirilgan:

| Funksiya | Tavsifi |
| :--- | :--- |
| `sin()` | Sinusni elementlar bo'yicha hisoblaydi. |
| `cos()` | Kosinusni elementlar bo'yicha hisoblaydi. |
| `tan()` | Tangensni elementlar bo'yicha hisoblaydi. |
| `arcsin()` | Arksinus (teskari sinus) amali. |
| `arccos()` | Arkkosinus (teskari kosinus) amali. |
| `arctan()` | Arktangens (teskari tangens) amali. |
| `arctan2()` | Ikki argumentli arktangens (kvadrantni to'g'ri tanlagan holda). |
| `degrees()` / `rad2deg()` | Burchaklarni radiandan darajaga o'tkazadi. |
| `radians()` / `deg2rad()` | Burchaklarni darajadan radianga o'tkazadi. |
| `hypot()` | To'g'ri burchakli uchburchak katetlari berilganda gipotenuzani hisoblaydi. |
| `unwrap()` | Qiymatlar orasidagi farqni $2\pi$ to'ldiruvchisi bilan o'zgartirish orqali "yoyish". |



> **Eslatma:** NumPy-dagi standart trigonometrik funksiyalar burchakni darajada emas, balki **radianda** qabul qiladi. Agar sizda burchaklar darajada bo'lsa, avval ularni `radians()` funksiyasi yordamida radianga o'tkazib olishingiz lozim.

### **Sinus funksiyasi (Sine Function)**

Sinus funksiyasi birlik aylanadagi nuqtaning berilgan burchakka (radianlarda) mos keluvchi **y-koordinatasini** qaytaradi.

In [20]:
# Burchaklar ro'yxati (radianlarda)
arr = [0, np.pi/2, np.pi/3, np.pi]
s1 = np.sin(arr)

print("Sinus qiymatlari:\n", s1)

Sinus qiymatlari:
 [0.00000000e+00 1.00000000e+00 8.66025404e-01 1.22464680e-16]


**Izoh:**
* `arr` o'zgaruvchisida burchaklar radian o'lchov birligida saqlanadi.
* `np.sin(arr)` funksiyasi massivdagi har bir burchak uchun sinus qiymatlarini elementlar bo'yicha (element-wise) hisoblaydi.


### **Kosinus funksiyasi (Cosine Function)**

Kosinus funksiyasi birlik aylanadagi nuqtaning berilgan burchakka (radianlarda) mos keluvchi **x-koordinatasini** qaytaradi.

In [21]:
# Burchaklar ro'yxati (radianlarda)
arr = [0, np.pi/2, np.pi/3, np.pi]
c1 = np.cos(arr)

print("Kosinus qiymatlari:\n", c1)

Kosinus qiymatlari:
 [ 1.000000e+00  6.123234e-17  5.000000e-01 -1.000000e+00]


**Izoh:**
* `np.cos(arr)` barcha burchaklar uchun kosinus qiymatlarini elementlar bo'yicha hisoblaydi.

### **Muhim eslatma:**
Natijalardagi `1.22e-16` yoki `6.12e-17` kabi juda kichik sonlar matematik jihatdan **0** ga teng deb hisoblanadi. Bu kompyuterda $\pi$ sonini hisoblashdagi juda kichik o'nli xatoliklar tufayli yuzaga keladi.

### **Giperbolik funksiyalar (Hyperbolic Functions)**

Giperbolik funksiyalar giperbola asosidagi matematik amallarni bajarish uchun ishlatiladi. Ular ko'pincha fizik jarayonlarni modellashtirishda, masalan, osilib turgan simning shaklini (zanjir chizig'i) hisoblashda qo'llaniladi.



| Funksiya | Tavsifi |
| :--- | :--- |
| `sinh()` | Giperbolik sinusni hisoblaydi. |
| `cosh()` | Giperbolik kosinusni hisoblaydi. |
| `tanh()` | Giperbolik tangensni elementlar bo'yicha hisoblaydi. |
| `arcsinh()` | Teskari giperbolik sinusni elementlar bo'yicha hisoblaydi. |
| `arccosh()` | Teskari giperbolik kosinusni elementlar bo'yicha hisoblaydi. |
| `arctanh()` | Teskari giperbolik tangensni elementlar bo'yicha hisoblaydi. |

### **Giperbolik sinus (Hyperbolic Sine)**

Giperbolik sinus (`sinh`) funksiyasi sinus funksiyasining giperbolaga asoslangan analogi bo'lib, u quyidagi formula orqali aniqlanadi:
$$\sinh(x) = \frac{e^x - e^{-x}}{2}$$



**Kod misoli:**

In [22]:
# Qiymatlar ro'yxati
arr = [0, np.pi/2, np.pi/3, np.pi]

# Giperbolik sinusni hisoblash
sh = np.sinh(arr)
print("Giperbolik sinus qiymatlari:\n", sh)

Giperbolik sinus qiymatlari:
 [ 0.          2.3012989   1.24936705 11.54873936]


**Izoh:**
`np.sinh(arr)` funksiyasi massivdagi har bir qiymat uchun giperbolik sinusni elementlar bo'yicha (element-wise) hisoblab chiqadi. Oddiy sinusdan farqli o'laroq, bu funksiya burchaklar bilan emas, balki eksponental o'sishga asoslangan qiymatlar bilan ishlaydi.

### **Yaxlitlash funksiyalari (Rounding Functions)**

Yaxlitlash funksiyalari sonlarni eng yaqin butun songacha yoki ko'rsatilgan o'nlik belgilargacha yaxlitlash uchun ishlatiladi. NumPy ma'lumotlar bilan ishlashda va hisoblash jarayonlarini optimallashtirishda poydevor vazifasini o'taydi.

Quyida NumPy-dagi asosiy yaxlitlash funksiyalari va ularning tavsifi keltirilgan:

| Funksiya | Tavsifi (Izoh) |
| :--- | :--- |
| `rint()` | Eng yaqin butun songa qarab yaxlitlaydi. |
| `fix()` | Nolga qarab eng yaqin butun songa yaxlitlaydi. |
| `floor()` | Kiruvchi qiymatning "pol" (pastki butun son) qismini elementlar bo'yicha qaytaradi. |
| `ceil()` | Kiruvchi qiymatning "shift" (yuqori butun son) qismini elementlar bo'yicha qaytaradi. |
| `trunc()` | Kiruvchi qiymatning kesilgan (faqat butun) qismini elementlar bo'yicha qaytaradi. |

### **Funksiyalarning ishlashiga misollar**

Ushbu funksiyalar massiv elementlari ustida quyidagicha amallarni bajaradi:

* `np.floor()`: Har doim kichik butun songa qarab siljiydi (masalan, `3.7` $\rightarrow$ `3.0`, `-3.7` $\rightarrow$ `-4.0`).
* `np.ceil()`: Har doim katta butun songa qarab siljiydi (masalan, `3.1` $\rightarrow$ `4.0`, `-3.1` $\rightarrow$ `-3.0`).
* `np.trunc() va np.fix()`: Sonning verguldan keyingi qismini shunchaki tashlab yuboradi (masalan, `3.9` $\rightarrow$ `3.0`, `-3.9` $\rightarrow$ `-3.0`).



Bu funksiyalar yirik, ko'p o'lchamli massivlar va matritsalar bilan ishlashda ma'lumotlarni aniqlik darajasini boshqarish uchun juda muhimdir

**`rint() funksiyasi`**

`np.rint()` funksiyasi massivdagi har bir elementni eng yaqin butun songacha yaxlitlaydi va yaxlitlangan qiymatlardan iborat yangi massiv qaytaradi.



**Misol:**

In [23]:
arr = [1.2, 2.7, -3.4, -4.6]
print(np.rint(arr))

[ 1.  3. -3. -5.]


### **Eksponentalar va logarifmlar funksiyalari**

NumPy eksponenta, logarifm va darajaga ko'tarish amallarini har bir element uchun alohida (element-wise) bajarishni qo'llab-quvvatlaydi.

| Funksiya | Tavsifi |
| :--- | :--- |
| `np.exp()` | Har bir element uchun $e^x$ (eksponenta) ni hisoblaydi. |
| `expm1()` | Massivdagi barcha elementlar uchun $e^x - 1$ ni hisoblaydi. |
| `exp2()` | Kiruvchi massivdagi barcha $p$ elementlari uchun $2^p$ ni hisoblaydi. |
| `log10()` | Kiruvchi massivning 10 asosli logarifmini qaytaradi. |
| `log2()` | $x$ ning 2 asosli logarifmi. |
| `log1p()` | Kiruvchi massivning har bir elementi uchun natural logarifm plus bir, ya'ni $\ln(1+x)$ ni qaytaradi. |
| `logaddexp()` | Kirishlarning eksponentalari yig'indisining logarifmi: $\ln(e^{x1} + e^{x2})$. |
| `logaddexp2()` | Kirishlarning 2 asosli eksponentalari yig'indisining 2 asosli logarifmi. |

### **Amaliy misol (exp va log):**

In [28]:
arr = [1, 2, 3]
# Eksponenta (e^1, e^2, e^3)
print("Eksponenta:", np.exp(arr))

# Natural logarifm (ln)
arr2 = [1, 10, 100]
print("10 asosli logarifm:", np.log10(arr2))

Eksponenta: [ 2.71828183  7.3890561  20.08553692]
10 asosli logarifm: [0. 1. 2.]


### **Eksponenta (Exponential)**

`np.exp()` funksiyasi massivdagi har bir element uchun $e^x$ qiymatini hisoblaydi. Bu yerda $e$ soni natural logarifm asosi bo‘lib, taxminan **2.718** ga teng.

**Misol:**

In [25]:
arr = [1, 3, 5]
print(np.exp(arr))

[  2.71828183  20.08553692 148.4131591 ]


### **Natural logarifm (Natural Logarithm)**

`np.log()` funksiyasi massivdagi har bir elementning $e$ asosiga ko‘ra logarifmini (natural logarifm) hisoblaydi. 

Matematikada bu $\ln(x)$ ko‘rinishida ham yoziladi.

**Misol:**

In [26]:
arr = [1, 3, 5, 256]
print(np.log(arr))

[0.         1.09861229 1.60943791 5.54517744]


**Izoh:**
* `np.log(1)` natijasi **0.0** bo‘ladi, chunki $e^0 = 1$.
* Funksiya har bir element ustida mustaqil ravishda (element-wise) amal bajaradi.

### **Arifmetik funksiyalar (Arithmetic Functions)**

Arifmetik funksiyalar Python-da qo'shish, ayirish, ko'paytirish va bo'lish kabi asosiy matematik amallarni bajaradi. NumPy ushbu amallarni massivlar ustida elementlar bo'yicha (element-wise) tezkor amalga oshirish imkonini beradi.



Quyidagi jadvalda NumPy-dagi asosiy arifmetik funksiyalar va ularning tavsifi keltirilgan:

| Funksiya | Tavsifi |
| :--- | :--- |
| `add()` | Argumentlarni elementlar bo'yicha qo'shadi. |
| `subtract()` | Argumentlarni elementlar bo'yicha ayiradi. |
| `multiply()` | Argumentlarni elementlar bo'yicha ko'paytiradi. |
| `divide()` | Massivlar yoki sonlar orasida elementlar bo'yicha bo'lishni amalga oshiradi. |
| `true_divide()` | Kirishlarning haqiqiy (o'nli kasr) bo'linmasini qaytaradi. |
| `floor_divide()` | Bo'linmaning natijasidan kichik yoki teng bo'lgan eng katta butun sonni qaytaradi. |
| `power()` | Birinchi massiv elementlarini ikkinchi massivdagi mos darajalarga ko'taradi. |
| `float_power()` | Birinchi massiv elementlarini haqiqiy son ko'rinishida darajaga ko'taradi. |
| `mod()` / `remainder()` | Bo'lishdan hosil bo'lgan qoldiqni qaytaradi. |
| `divmod()` | Bir vaqtning o'zida bo'linma (butun qism) va qoldiqni qaytaradi. |
| `reciprocal()` | Har bir element uchun teskari qiymatni (1/x) qaytaradi. |
| `positive()` | Sonning musbat qiymatini (o'zini) qaytaradi. |
| `negative()` | Sonning manfiy qiymatini qaytaradi. |



### **Amaliy misollar:**

* **Qo'shish va ayirish:** Ikki massivning mos indekslaridagi sonlarni birlashtiradi yoki biridan ikkinchisini ayiradi.
* **Bo'lish turlari:** `true_divide` oddiy bo'lish kabi natija (masalan, 7/2 = 3.5) bersa, `floor_divide` faqat butun qismini (masalan, 7/2 = 3) oladi.
* **Qoldiqni topish:** `mod()` funksiyasi matematik qoldiqni hisoblab beradi.

Ushbu amallar NumPy-ning yuqori samaradorligi tufayli juda katta hajmli ma'lumotlar to'plami bilan ishlashda ham bir zumda bajariladi.

### **Divide funksiyasi (Divide Function)**

`numpy.divide(arr1, arr2)` funksiyasi birinchi massiv elementlarini ikkinchi massivning mos elementlariga elementlar bo'yicha (**element-wise**) bo'lishni amalga oshiradi. Bu funksiya har doim "haqiqiy bo'lish" (true division) natijasini qaytaradi, ya'ni natija butun son bo'lsa ham, u suzuvchi nuqtali (**float**) ko'rinishida bo'ladi.

**Misol:**

In [27]:
# Kirish massivlari
arr1 = [2, 27, 2, 21, 23]
arr2 = [2, 3, 4, 5, 6]

# Elementlar bo'yicha bo'lish
natija = np.divide(arr1, arr2)

print(natija)

[1.         9.         0.5        4.2        3.83333333]


### **Mantiqiy tahlil va tushuntirish:**



Funksiya massivlarni indeksma-indeks quyidagicha hisoblaydi:
* $2 / 2 = 1.0$
* $27 / 3 = 9.0$
* $2 / 4 = 0.5$
* $21 / 5 = 4.2$
* $23 / 6 = 3.83333333$

**Muhim jihatlari:**
* **Massivlar shakli:** `arr1` va `arr2` massivlari bir xil o'lchamda bo'lishi yoki **Broadcasting** (translyatsiya) qoidalariga mos kelishi kerak.
* **0 ga bo'lish:** Agar `arr2` tarkibida nol bo'lsa, NumPy xatolik bermaydi, balki ogohlantirish (warning) berib, natijada `inf` (infinity - cheksizlik) yoki `nan` (not a number) qiymatlarini qaytaradi.
* **Soddaligi:** NumPy-da `np.divide(arr1, arr2)` o'rniga oddiygina `arr1 / arr2` operatoridan foydalanish ham xuddi shu natijani beradi.

### **Kompleks sonlar bilan ishlash funksiyalari**

NumPy kompleks sonlar (haqiqiy va mavhum qismlardan iborat sonlar) ustida turli amallarni bajarish va ularning turini tekshirish uchun maxsus funksiyalarni taqdim etadi.

**`isreal()` funksiyasi**

`numpy.isreal()` funksiyasi massiv elementlarining **haqiqiy son** ekanligini elementlar bo‘yicha tekshiradi. Agar elementning mavhum qismi (imaginary part) nolga teng bo‘lsa, u haqiqiy son hisoblanadi va `True` qaytaradi.



**Misol:**
* **Kompleks massiv bilan:** `[1+1j, 0j]` kiritilganda, birinchi element mavhum qismga ega bo‘lgani uchun `False`, ikkinchi element (`0j`) esa mavhum qismi 0 bo‘lgani uchun `True` natijasini beradi.
* **Oddiy massiv bilan:** `[1, 0]` kiritilganda, barcha elementlar haqiqiy bo‘lgani uchun `[True, True]` natijasi hosil bo‘ladi.

**`conj()` funksiyasi**

`numpy.conj(x)` funksiyasi kompleks sonning **qo‘shmasini (conjugate)** hisoblaydi. Qo‘shma sonni hosil qilish uchun sonning faqat **mavhum qismining ishorasi** o‘zgartiriladi (musbat bo‘lsa manfiyga, manfiy bo‘lsa musbatga).



**Misol:**
* **2+4j** sonining qo‘shmasi: Mavhum qism ishorasi o‘zgarib, **2-4j** bo‘ladi.
* **5-8j** sonining qo‘shmasi: Mavhum qism ishorasi o‘zgarib, **5+8j** bo‘ladi.

### **Qisqacha jadval**

| Funksiya | Tavsifi | Natija turi |
| :--- | :--- | :--- |
| `isreal()` | Elementning haqiqiy son ekanligini tekshiradi. | Boolean (`True`/`False`) |
| `conj()` | Kompleks sonning qo‘shmasini qaytaradi. | Kompleks son |
| `imag()` | Sonning faqat mavhum qismini qaytaradi. | Haqiqiy son |
| `real()` | Sonning faqat haqiqiy qismini qaytaradi. | Haqiqiy son |

### **Maxsus funksiyalar (Special functions)**

Maxsus funksiyalar konvolyutsiya (o'rash), elementlar bo'yicha hisob-kitoblar, interpolyatsiya hamda **NaN** (son emas) yoki kompleks qiymatlar bilan ishlash kabi murakkab matematik va sonli amallarni bajarish uchun mo'ljallangan.



Quyidagi jadvalda NumPy-dagi asosiy maxsus funksiyalar va ularning tavsifi keltirilgan:

| Funksiya | Tavsifi |
| :--- | :--- |
| `convolve()` | Ikki bir o'lchamli ketma-ketlikning diskret, chiziqli konvolyutsiyasini (o'rashini) qaytaradi. |
| `sqrt()` | Massivning manfiy bo'lmagan kvadrat ildizini elementlar bo'yicha hisoblaydi. |
| `square()` | Kiruvchi elementlarning kvadratini qaytaradi. |
| `absolute()` / `fabs()` | Elementlarning modulini (absolyut qiymatini) hisoblaydi. |
| `sign()` | Sonning ishorasini ko'rsatuvchi qiymatni qaytaradi (-1, 0 yoki 1). |
| `interp()` | Bir o'lchamli chiziqli interpolyatsiya. |
| `maximum()` / `minimum()` | Ikki massivning mos elementlari orasidagi maksimal yoki minimal qiymatni tanlaydi. |
| `real_if_close()` | Agar kompleks sonning mavhum qismi nolga yaqin bo'lsa, uni haqiqiy songa aylantiradi. |
| `nan_to_num()` | `NaN` qiymatlarni nolga, cheksizlikni esa katta chekli sonlarga almashtiradi. |
| `heaviside()` | Xevisayd (Heaviside) pog'onali funksiyasini hisoblaydi. |
| `cbrt()` | Massivning har bir elementi uchun kub ildizni hisoblaydi. |
| `clip()` | Massiv qiymatlarini belgilangan minimum va maksimum oraliq bilan chegaralaydi. |


### **Misollar**

* `sqrt() va square():` Masalan, `[4, 9]` massivining `sqrt` natijasi `[2.0, 3.0]` bo'ladi. `[2, 3]` ning `square` (kvadrat) natijasi esa `[4, 9]` bo'ladi.
* `absolute():` Manfiy sonlarni musbatga aylantiradi. Misol: `-5` natijasi `5` bo'ladi.
* `maximum():` Ikki massivni solishtiradi. Masalan, `[10, 2]` va `[5, 8]` massivlari uchun `maximum` natijasi `[10, 8]` bo'ladi.
* `clip():` Qiymatlarni "qirqib" qo'yadi. Masalan, massivni `[0, 10]` oralig'iga `clip` qilsangiz, `-5` soni `0` ga, `15` soni esa `10` ga aylanadi.
* `nan_to_num():` Ma'lumotlar to'plamidagi xatoliklarni (NaN) tozalashda ishlatiladi. Masalan, `[1, np.nan]` natijasi `[1, 0]` bo'ladi.



Bu funksiyalar ma'lumotlarni qayta ishlash va hisoblash jarayonlarini optimallashtirishda, ayniqsa mashinali o'rganish va ilmiy tadqiqotlarda juda muhim vosita hisoblanadi.

### **Kub ildiz (Cube Root)**

`numpy.cbrt(x)` funksiyasi massivning har bir elementi uchun kub ildizni hisoblaydi. `sqrt()` (kvadrat ildiz) funksiyasidan farqli o'laroq, `cbrt()` manfiy sonlar bilan ham ishlay oladi.

**Misol:**
Massiv elementlari `[1, 27000, 64, -1000]` bo'lganda, natija quyidagicha bo'ladi:
* $1$ ning kub ildizi: **1.0**
* $27000$ ning kub ildizi: **30.0**
* $64$ ning kub ildizi: **4.0**
* $-1000$ ning kub ildizi: **-10.0**

### **`clip()` funksiyasi**

`numpy.clip(x, a_min, a_max)` funksiyasi massivdagi qiymatlarni belgilangan minimal va maksimal oraliq bilan chegaralash (qirqish) uchun ishlatiladi.

**Misol:**
Agar bizda `[1, 2, 3, 4, 5, 6, 7, 8]` massivi bo'lsa va biz uni **2** (min) hamda **6** (max) oralig'ida chegaralasak:
* **2 dan kichik** bo'lgan har qanday son (masalan, 1) **2** ga aylanadi.
* **6 dan katta** bo'lgan har qanday son (masalan, 7, 8) **6** ga aylanadi.
* **2 va 6 oralig'idagi** sonlar o'zgarishsiz qoladi.

**Natija:** `[2, 2, 3, 4, 5, 6, 6, 6]`

Ushbu funksiya ma'lumotlardagi keskin chiquvchi qiymatlarni (outliers) jilovlashda juda foydali hisoblanadi.

### 1.3.3. Satrli amallar (String Operations)

NumPy kutubxonasining matnli funksiyalari **numpy.char** moduliga tegishli bo‘lib, massivlar ustida elementlar bo‘yicha (**element-wise**) amallarni bajarishga mo‘ljallangan. Ushbu funksiyalar matnli ma’lumotlarni samarali boshqarish va ularga ishlov berishda yordam beradi.